torch/_dynamo/utils.py

```diff
@@ -1320,6 +1297,60 @@ def add_compilation_metrics_to_chromium(c: CompilationMetrics) -> None:
     )
 
 
+class RuntimeCompileContext:
+    """
+    For tracking compile-time information that we want available at runtime,
+    e.g., to allow us to query and log at runtime the CompileId active when
+    a module was compiled.
+    """
+
+    @staticmethod
+    def get() -> RuntimeCompileContext:
+        assert _TLS.runtime_compile_context is not None
+        return _TLS.runtime_compile_context
+
+    @staticmethod
+    def try_get() -> Optional[RuntimeCompileContext]:
+        return getattr(_TLS, "runtime_compile_context", None)
+
+    def __init__(self, trace_id: Optional[TraceId]):
+        self.trace_id = trace_id
+
+    @staticmethod
+    def current_trace_id() -> Optional[TraceId]:
+        ctx = RuntimeCompileContext.try_get()
+        return None if ctx is None else ctx.trace_id
+
+    @staticmethod
+    def current_compile_id() -> Optional[CompileId]:
+        trace_id = RuntimeCompileContext.current_trace_id()
+        return None if trace_id is None else trace_id.compile_id
+
+
+_TLS = threading.local()
+
+
+@contextmanager
+def runtime_compile_context(context: Optional[RuntimeCompileContext]):
```

Comment:
something like this

```suggestion
def runtime_compile_context(context: Optional[RuntimeCompileContext]):
    assert CompileContext.try_get() is None
```

and asserting no runtime compile context from compile_context. This runs in circular import, so we may want to consolidate both into the same file and TLS